markdown_help.go

```diff
@@ -0,0 +1,161 @@
+package baker
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+)
+
+// GenerateMarkdownHelp generates markdown-formatted textual help for a Baker
+// component from its description structure. Markdown is written into w.
+func GenerateMarkdownHelp(w io.Writer, desc interface{}) error {
+	if desc == nil {
+		return fmt.Errorf("can't generate markdown help for a nil interface")
+	}
+
+	if reflect.TypeOf(desc).Kind() == reflect.Ptr {
+		// dereference pointer
+		desc = reflect.ValueOf(desc).Elem().Interface()
+	}
+
+	switch d := desc.(type) {
+	case InputDesc:
+		doc, err := newInputDoc(d)
+		if err != nil {
+			return err
+		}
+		genInputMarkdown(w, doc)
+	case FilterDesc:
+		doc, err := newFilterDoc(d)
+		if err != nil {
+			return err
+		}
+		genFilterMarkdown(w, doc)
+	case OutputDesc:
+		doc, err := newOutputDoc(d)
+		if err != nil {
+			return err
+		}
+		genOutputMarkdown(w, doc)
+	case UploadDesc:
+		doc, err := newUploadDoc(d)
+		if err != nil {
+			return err
+		}
+		genUploadMarkdown(w, doc)
+	case MetricsDesc:
+		doc, err := newMetricsDoc(d)
+		if err != nil {
+			return err
+		}
+		genMetricsMarkdown(w, doc)
+	default:
+		return fmt.Errorf("can't generate markdown help, unsupported type %T", desc)
+	}
+
+	return nil
+}
+
+func breakAfterDots(s string) string {
```

Comment:
in markdown, newline requires two spaces before a `\n` if we don't want to leave a blank line or `\n\n` otherwise. A single `\n` is rendered as a simple space without newline.

Reply:
uh?!  a single `\n` is rendered as a `\n` in markdown.  
The objective here is to add a carriage return (i.e just go to the next line) and this is what that does. 
What you're saying with the 2 spaces is `\n\n` is if you want to force a line jump on a blank line.
Here we're not dealing with blank lines, but line with text, so adding `\n` does what is expected.
Don't you agree?

Reply:
no really, markdown specs says that 2 spaces + `\n` correspond to a line break :) https://www.markdownguide.org/basic-syntax/#line-breaks
most markdown renderer also 
render a single `\n` as a line break, 
but many others, including GH, don't

Reply:
while 2 blank lines create a paragraph https://www.markdownguide.org/basic-syntax/#paragraphs-1